[section#except.terminate
    [:en] The [`std::terminate] function
    [:zh_CN] [`std::terminate] 函数
]

[para]
[%[`terminate]]
[:en]
In some situations exception handling must be abandoned for less subtle error
handling techniques.
[:zh_CN]
某些情况下必须放弃异常处理，而采用不这么微妙的错误处理技术。

[enter:note]
[:en]
These situations are:
[:zh_CN]
这些情况有：

[:en]
[%[`terminate][!called]]
[:zh_CN]
[%[`terminate][!调用]]
[list]
[item]
[:en] when the exception handling mechanism, after completing the initialization
of the exception object but before activation of a handler for the exception
([#except.throw]), calls a function that exits via an exception, or
[:zh_CN] 当对异常对象的初始化完成之后但在该异常的某个处理器开始活跃之前
（[#except.throw]），异常处理机制调用了某个以异常退出的函数时，或者

[item]
[:en] when the exception handling mechanism cannot find a handler for a thrown
exception ([#except.handle]), or
[:zh_CN] 当异常处理机制无法为所抛出的异常找到处理器时（[#except.handle]），或者

[item]
[:en] when the search for a handler ([#except.handle]) encounters the outermost
block of a function with a non-throwing exception specification
([#except.spec]), or
[:zh_CN] 当搜索处理器（[#except.handle]）时遇到一个函数的最外层块，而它带有
无抛出异常说明（[#except.spec]）时，或者

[item]
[:en] when the destruction of an object during stack unwinding ([#except.ctor])
terminates by throwing an exception, or
[:zh_CN] 当在栈解退（[#except.ctor]）过程中某对象的销毁过程以抛出异常退出时，
或者

[item]
[:en] when initialization of a non-local variable with static or thread storage
duration ([#basic.start.dynamic]) exits via an exception, or
[:zh_CN] 当带有静态或线程存储期的非局部变量的初始化过程（[#basic.start.dynamic]）
以异常退出时，或者

[item]
[:en] when destruction of an object with static or thread storage duration exits
via an exception ([#basic.start.term]), or
[:zh_CN] 当带有静态或线程存储期的对象的销毁过程以异常退出时
（[#basic.start.term]），或者

[item]
[:en] when execution of a function registered with [`std::atexit] or
[`std::at_quick_exit] exits via an exception ([#support.start.term]), or
[:zh_CN] 当用 [`std::atexit] 或 [`std::at_quick_exit] 登记的函数的执行以异常退出
时（[#support.start.term]），或者

[item]
[:en] when evaluation of the predicate of a contract ([#dcl.attr.contract])
exits via an exception, or
[:zh_CN] 当某个契约的谓词的求值（[#dcl.attr.contract]）以异常退出时，或者

[item]
[:en] when the violation handler invoked for a failed contract condition check
([#dcl.attr.contract]) on a [`noexcept] function exits via an exception, or
[:zh_CN] 当为某个 [`noexcept] 函数所进行的失败的谓词条件检查
（[#dcl.attr.contract]）所调用的违例处理器以异常退出时，或者

[item]
[:en] when the violation handler has completed after a failed contract check and
the continuation mode is [$off], or
[:zh_CN] 当对失败的契约检查之后调用的违例处理器完成执行，且继续模式为 [$off] 时，
或者

[item]
[:en] when a [~throw-expression] ([#expr.throw]) with no operand attempts to
rethrow an exception and no exception is being handled ([#except.throw]), or
[:zh_CN] 当没有操作数的 [~throw-表达式]（[#expr.throw]）试图重新抛出异常，而没有
正在被处理的异常时（[#except.throw]），或者

[item]
[:en] when the function [`std::nested_exception::rethrow_nested] is called for
an object that has captured no exception ([#except.nested]), or
[:zh_CN] 当为某个并未捕获异常的对象调用 [`std::nested_exception::rethrow_nested]
函数时（[#except.nested]），或者

[item]
[:en] when execution of the initial function of a thread exits via an exception
([#thread.thread.constr]), or
[:zh_CN] 当线程的起始函数以异常退出时（[#thread.thread.constr]），或者

[item]
[:en] for a parallel algorithm whose [`ExecutionPolicy] specifies such behavior
([#execpol.seq], [#execpol.par], [#execpol.parunseq]), when execution of an
element access function ([#algorithms.parallel.defns]) of the parallel algorithm
exits via an exception ([#algorithms.parallel.exceptions]), or
[:zh_CN] 对于 [`ExecutionPolicy] 指定了这种行为（[#execpol.seq]，[#execpol.par]，
[#execpol.parunseq]）的并行算法，当该并行算法的元素访问函数
（[#algorithms.parallel.defns]）的执行以异常退出时
（[#algorithms.parallel.exceptions]），或者

[item]
[:en] when the destructor or the move assignment operator is invoked on an
object of type [`std::thread] that refers to a joinable thread
([#thread.thread.destr], [#thread.thread.assign]), or
[:zh_CN] 当为某个代表可会合线程的 [`std::thread] 类型的对象调用其析构函数或移动
赋值运算符时（[#thread.thread.destr]，[#thread.thread.assign]），或者

[item]
[:en] when a call to a [`wait()], [`wait_until()], or [`wait_for()] function on
a condition variable ([#thread.condition.condvar],
[#thread.condition.condvarany]) fails to meet a postcondition.
[:zh_CN] 当对某个条件变量（[#thread.condition.condvar]，
[#thread.condition.condvarany]）所调用的 [`wait()]，[`wait_until()]，或
[`wait_for()] 函数无法满足其后条件时。

[list:end]

[exit:note]

[para]
[%[`terminate]]
[:en]
In such cases, the function [`std::terminate] is called ([#exception.terminate]).
[:zh_CN]
在这些情况下，函数 [`std::terminate] 将被调用（[#exception.terminate]）。

[:en]
In the situation where no matching handler is found, it is [%@impldef stack
unwinding before call to [`std::terminate]] implementation-defined whether or
not the stack is unwound before [`std::terminate] is called.
[:zh_CN]
在没有找到匹配处理器的情况中，是否于调用 [`std::terminate] 之前进行栈解退是
[%@impldef 在调用 [`std::terminate] 之前进行栈解退]由实现定义的。

[:en]
In the situation where the search for a handler ([#except.handle]) encounters
the outermost block of a function with a non-throwing exception specification
([#except.spec]), it is [%@impldef whether stack is unwound before calling the
function [`std::terminate] when a [`noexcept] specification is violated]
implementation-defined whether the stack is unwound, unwound partially, or not
unwound at all before the function [`std::terminate] is called.
[:zh_CN]
在搜索处理器（[#except.handle]）时遇到某个函数的最外层块，而它带有无抛出异常说明
（[#except.spec]）的情况中，在调用函数 [`std::terminate] 之前，栈是否完成解退，
是部分进行了解退，还是完全不进行解退，是[%@impldef 当违反 [`noexcept] 说明时，
调用函数 [`std::terminate] 之前是否进行栈解退]由实现定义的。

[:en]
In all other situations, the stack shall not be unwound before the function
[`std::terminate] is called.
[:zh_CN]
在所有其他情况下，调用函数 [`std::terminate] 之前不应进行栈解退。

[:en]
An implementation is not permitted to finish stack unwinding prematurely based
on a determination that the unwind process will eventually cause a call to the
function [`std::terminate].
[:zh_CN]
不允许实现基于确定栈解退过程最终将导致调用函数 [`std::terminate] 来提前完成
栈解退。
